src/query_plan/query_planning_traversal.rs

```diff
@@ -432,10 +434,50 @@ impl<'a> QueryPlanningTraversal<'a> {
 
     fn selection_set_is_fully_local_from_all_nodes(
         &self,
-        _selection: &NormalizedSelectionSet,
-        _nodes: &IndexSet<NodeIndex>,
+        selection: &NormalizedSelectionSet,
+        nodes: &IndexSet<NodeIndex>,
     ) -> Result<bool, FederationError> {
-        todo!()
+        // To guarantee that the selection is fully local from the provided vertex/type, we must have:
+        // - no edge crossing subgraphs from that vertex.
+        // - the type must be compositeType (mostly just ensuring the selection make sense).
+        // - everything in the selection must be avaiable in the type (which `rebaseOn` essentially validates).
+        // - the selection must not "type-cast" into any abstract type that has inconsistent runtimes acrosse subgraphs. The reason for the
+        //   later condition is that `selection` is originally a supergraph selection, but that we're looking to apply "as-is" to a subgraph.
+        //   But suppose it has a `... on I` where `I` is an interface. Then it's possible that `I` includes "more" types in the supergraph
+        //   than in the subgraph, and so we might have to type-explode it. If so, we cannot use the selection "as-is".
+        // PORT_NOTE: The JS code performs the last check lazily. Instead of that, this check is
+        // skipped if `nodes` is empty.
+        if !nodes.is_empty()
+            && selection.selections.keys().any(|k| match k {
+                NormalizedSelectionKey::InlineFragment {
+                    type_condition: Some(name),
+                    ..
+                } => self
+                    .parameters
+                    .abstract_types_with_inconsistent_runtime_types
+                    .iter()
+                    .any(|ty| ty.type_name() == name),
```

Comment:
hm, I think this is the right thing to do right now, but I don't love that we aren't using the Set-ness of `abstract_types_with_inconsistent_runtime_types`. maybe its type needs to be less strict, so we can use `.contains()` even when we don't have an `AbstractTypeDefinitionPosition` like here.